gobblin-config-management/src/main/java/gobblin/config/configstore/impl/BaseHdfsConfigStore.java

```diff
@@ -0,0 +1,214 @@
+      LOG.error("can not find the version " + version + " error:  " + ioe.getMessage(), ioe);
+      throw new VersionDoesNotExistException(ioe.getMessage());
+    } 
+  }
+
+  @Override
+  public Collection<URI> getChildren(URI uri, String version) {
+    if (uri == null)
+      return null;
+    
+    Path self = getPath(uri, version) ;
+    
+    try {
+      if (this.fs.isFile(self)) {
+        return Collections.emptyList();
+      }
+
+      FileStatus[] fileStatus = this.fs.listStatus(self);
+      if (fileStatus == null || fileStatus.length == 0) {
+        return Collections.emptyList();
+      }
+
+      List<URI> res = new ArrayList<URI>();
+      for (FileStatus f : fileStatus) {
+        // valid node should be a directory
+        if (!f.isDir()) {
+          continue;
+        }
+        try {
+          res.add(new URI(this.getRelativePath(f.getPath(), version)));
+        } catch (URISyntaxException e) {
+          LOG.error(String.format("Got error when create URI for %s, exception %s", f.getPath(), e.getMessage()), e);
+          throw new RuntimeException(e);
+        }
+      }
+
+      return res;
+    } catch (IOException ioe) {
+      LOG.error(String.format("Got error when find children for %s, exception %s", self, ioe.getMessage()));
+      throw new RuntimeException(ioe);
+    }
+  }
+  
+  @Override
+  public Config getOwnConfig(URI uri, String version) {
+    if (uri == null)
+      return ConfigFactory.empty();
+
+    Closer closer = Closer.create();
+    Path self = getPath(uri, version);
+    Path configFile = new Path(self, CONFIG_FILE_NAME);
+    try {
+      if (!this.fs.isFile(configFile)) {
+        return ConfigFactory.empty();
+      }
+
+      FSDataInputStream configFileStream = closer.register(this.fs.open(configFile));
+      return ConfigFactory.parseReader(new InputStreamReader(configFileStream)).resolve();
+    }
+    catch (IOException ioe) {
+      LOG.error(String.format("Got error when get own config for %s, exception %s", self, ioe.getMessage()), ioe);
+      throw new RuntimeException(ioe);
+    }
+    finally {
+      try {
+        closer.close();
+      } catch (IOException e) {
+        LOG.error("Failed to close FsInputStream: " + e, e);
+      }
+    }
+  }
+
+  protected String getRelativePath(Path p, String version) {
+    String root = PathUtils.getPathWithoutSchemeAndAuthority(this.getVersionRoot(version)).toString();
+    String input = PathUtils.getPathWithoutSchemeAndAuthority(p).toString();
+
+    if (input.equals(root)) {
+      return "";
+    }
+    return input.substring(root.length()+1);
+  }
+  
+  protected static final boolean isRootURI(URI uri) {
+    if (uri == null)
+      return false;
+
+    return uri.toString().length()==0;
+  }
+  
+  protected final Path getPath(URI uri, String version){
+    Path self ;
+    if(isRootURI(uri)){
+      self = this.getVersionRoot(version);
+    }else{
+      self = new Path(this.getVersionRoot(version), uri.toString());
```

Comment:
`URI.toString()` is a relatively expensive operation, and `Path` uses URI's internally either way. It is more efficient to avoid `toString` as much as possible. For example `new Path(this.getVersionRoot(version), new Path(uri))`